Allow pulling all nonlocal 'a' coefficient data

lindsayad opened the pull request:
Closes #24453

lindsayad committed:
Allow pulling all nonlocal 'a' coefficient data

Closes #24453

GiudGiud commented on modules/navier_stokes/src/userobjects/INSFVRhieChowInterpolator.C:
```diff
@@ -439,14 +440,28 @@ INSFVRhieChowInterpolator::finalize()
+  if (_pull_all_nonlocal)
+  {
+    for (const auto * const elem :
+         as_range(_mesh.active_not_local_elements_begin(), _mesh.active_not_local_elements_end()))
```
I m curious how we know an element is active if it s not local

same how do we know their processor id or the elem id?

lindsayad replied:
If any process has an element then libMesh ensures that the data in that element is complete and uniform across all processes

GiudGiud replied:
even if the mesh is distributed?

lindsayad replied:
correct although in that case many processes may not have a given element at all

GiudGiud replied:
so the process does not have it, so it doesnt show up in the "non_local" list?

feels like "non_local" is  more "replicated_non_local"

lindsayad replied:
To me its implicit in `elements` that the range will not have null/nonexistent elements

lindsayad replied:
And `ghosted_non_local` means the same to me as `non_local`

lindsayad replied:
But you can debate this name with @roystgnr haha

roystgnr replied:
That name is kinda grandfathered in.  Added by John on Sep 7 2007; I started "ParallelMesh" a couple weeks later.

GiudGiud replied:
> And ghosted_non_local means the same to me as non_local

non_local to me could non_local and (ghosted or non_ghosted). Just not local.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1544051/) on d9cd69b wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/24454/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1544069/) on d9cd69b wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>395ab6</th>
      <th colspan=4>#24454 d9cd69</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.36%</td>
      <td>85.36%</td>
      <td>+0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>88275</td>
      <td>88279</td>
      <td>+4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15139</td>
      <td>15135</td>
      <td>-4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24454/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24454/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>395ab6</th>
      <th colspan=4>#24454 d9cd69</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.86%</td>
      <td>83.84%</td>
      <td>-0.02%</td>
      <td>66.67%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>10799</td>
      <td>10800</td>
      <td>+1</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2079</td>
      <td>2082</td>
      <td>+3</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/24454/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/24454/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/24454/coverage/xfem)

</details>

## Warnings

- `navier_stokes` new line coverage rate 66.67% is less than the suggested 90.0%



This comment will be updated on new commits.

GiudGiud reviewed (changes requested):
- needs testing

- needs to be usable from the input file (so like a new parameter to the RC UO?)

- ideally, on a ghosting error on a coefficient, we would also tell a user to enable this.

lindsayad committed:
Move GhostEverything RM from test to framework

lindsayad committed:
Add assertion for remote elements in loopOverElemFaceInfo

lindsayad committed:
Address review

- Add user input parameter for pulling nonlocal 'a' data
- Add test
- Add some documentation

moosebuild commented:
Job [Conda (Rocky)](https://civet.inl.gov/job/1547776/) on 123c664 : invalidated by @lindsayad

lindsayad commented:
Thank you sir for the review!
